H5Dchunk_iter now passes offsets in units of dataset elements, fix #1419

mkitti opened the pull request:
This fixes #1419 by adding a `H5O_layout_chunk_t` pointer to `H5D_chunk_iter_ud_t` and using the `dim` and `ndim` fields.

This uses an statically sized array `hsize_t                    offset[H5O_LAYOUT_NDIMS]` which is passed to the user callback function.

`test/chunk_info.c` is modified to match the changes and is now consistent with `H5Dget_chunk_info`.

## Tasks
- [x] Update the documentation (H5Dpublic.h) to clarify that "logical coordinates of chunk" uses units of dataset elements

mkitti committed:
H5Dchunk_iter now passes chunk dimension scaled offsets, fix #1419

fortnern commented:
We should probably update the documentation (H5Dpublic.h> to clarify that "logical coordinates of chunk" uses units of dataset elements

derobins commented:
I'll approve and merge this once the documentation is updated.

mkitti committed:
Update docs for H5Dchunk_iter, H5Dget_chunk_info*

Modified description for `H5Dchunk_iter`, `H5Dget_chunk_info`, and `H5Dget_chunk_info_by_coord` to the following

 * offset          Logical position of the chunk’s first element in units of dataset elements
 * filter_mask  Bitmask indicating the filters used when the chunk was written
 * size             Chunk size in bytes, 0 if the chunk does not exist

mkitti commented:
We should regularize the description of the arguments. The descriptions of the arguments should be the same between `H5Dchunk_iter`, `H5Dget_chunk_info`, and `H5Dget_chunk_info_by_coord`. I revised the documentation for all three and regularized the changes so identical parameters have the same descriptions.

Also I changed `nbytes` to `size` since that has precedence. However, perhaps `nbytes` is a clearer name for this argument.

fortnern commented:
Looks good to me. Thanks!

derobins commented:
This also needs a note in RELEASE.txt, but I'll take care of that.

mkitti commented:
Thank you for the reviews. I will leave this branch alone now, and begin moving these changes into 1.12 via #1970 .

ZhaoMJ commented on src/H5Dchunk.c:
```diff
@@ -8066,13 +8067,20 @@ static int
+    unsigned                   ii; /* Match H5O_layout_chunk_t.ndims */
+
+    /* Similar to H5D__get_chunk_info */
+    for (ii = 0; ii < chunk->ndims; ii++)
```
I think this is off by one?  The `ndims` here is 1 larger than the rank I set. I also see in other places, `layout->ndims` is decreased by 1 to determine the number of dimensions.
https://github.com/HDFGroup/hdf5/blob/509fe962d42b3f02d356d2668b04712df06daf7f/src/H5Dearray.c#L1095-L1100

ZhaoMJ replied:
If you add an assertion here and run [this test](https://github.com/HDFGroup/hdf5/blob/509fe962d42b3f02d356d2668b04712df06daf7f/test/chunk_info.c#L1579), it's easy to see that `chunk->ndims` is 3 but we clearly expect the `offset` array to have two valid elements.

mkitti replied:
You are correct that there are ndims - 1 dimensions, but this is an internal implementation detail of the library. That specific loop comes from `H5D__get_chunk_info`.

What is the actual problem though? At worse, we do one more multiplication than needed while avoiding a subtraction and another variable on the stack.
